### PR TITLE
[TEST] wait until reindex tasks ready for rethrottle

### LINKED ISSUE #26192: [CI] org.elasticsearch.index.reindex.RethrottleTests with two tests failing
* Build failure link: https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+master+g1gc/3904/console
```
Tests with failures:
  - org.elasticsearch.index.reindex.RethrottleTests.testDeleteByQueryWithWorkers
  - org.elasticsearch.index.reindex.RethrottleTests.testDeleteByQuery
```

* Error message: 
```
ERROR   0.39s J1 | RethrottleTests.testDeleteByQueryWithWorkers <<< FAILURES!
   > Throwable #1: ElasticsearchException[Rethrottle of [kDDav6kbTqm-2S0E_VSGmA:163] failed]; nested: IllegalArgumentException[task [163] must be set as a child or parent];
   > 	at __randomizedtesting.SeedInfo.seed([B60C174C1CAD6D39:316B3A20264932DA]:0)
   > 	at org.elasticsearch.action.support.tasks.BaseTasksResponse.lambda$rethrowFailures$0(BaseTasksResponse.java:73)
   > 	at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:193)
   > 	at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1374)
   > 	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481)
   > 	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)
   > 	at java.util.stream.StreamSpliterators$WrappingSpliterator.forEachRemaining(StreamSpliterators.java:312)
   > 	at java.util.stream.Streams$ConcatSpliterator.forEachRemaining(Streams.java:743)
   > 	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481)
   > 	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)
   > 	at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:708)
   > 	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
   > 	at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:499)
   > 	at org.elasticsearch.action.support.tasks.BaseTasksResponse.rethrowFailures(BaseTasksResponse.java:74)
   > 	at org.elasticsearch.index.reindex.RethrottleTests.testCase(RethrottleTests.java:118)
   > 	at org.elasticsearch.index.reindex.RethrottleTests.testDeleteByQueryWithWorkers(RethrottleTests.java:71)
   > 	at java.lang.Thread.run(Thread.java:748)
   > Caused by: java.lang.IllegalArgumentException: task [163] must be set as a child or parent
   > 	at org.elasticsearch.index.reindex.TransportRethrottleAction.rethrottle(TransportRethrottleAction.java:73)
   > 	at org.elasticsearch.index.reindex.TransportRethrottleAction.taskOperation(TransportRethrottleAction.java:57)
   > 	at org.elasticsearch.index.reindex.TransportRethrottleAction.taskOperation(TransportRethrottleAction.java:43)
   > 	at org.elasticsearch.action.support.tasks.TransportTasksAction.nodeOperation(TransportTasksAction.java:152)
   > 	at org.elasticsearch.action.support.tasks.TransportTasksAction.access$1000(TransportTasksAction.java:67)
   > 	at org.elasticsearch.action.support.tasks.TransportTasksAction$NodeTransportHandler.messageReceived(TransportTasksAction.java:347)
   > 	at org.elasticsearch.action.support.tasks.TransportTasksAction$NodeTransportHandler.messageReceived(TransportTasksAction.java:343)
   > 	at org.elasticsearch.transport.TransportRequestHandler.messageReceived(TransportRequestHandler.java:30)
   > 	at org.elasticsearch.transport.RequestHandlerRegistry.processMessageReceived(RequestHandlerRegistry.java:66)
   > 	at org.elasticsearch.transport.TcpTransport$RequestHandler.doRun(TcpTransport.java:1526)
   > 	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:638)
   > 	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37)
   > 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
   > 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
   > 	... 1 more
```

```
FAILURE 0.73s J1 | RethrottleTests.testDeleteByQuery <<< FAILURES!
   > Throwable #1: java.lang.AssertionError: 
   > Expected: a collection with size a value less than <2>
   >      but: collection size <2> was equal to <2>
   > 	at __randomizedtesting.SeedInfo.seed([B60C174C1CAD6D39:B8C31267456212AF]:0)
   > 	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
   > 	at org.elasticsearch.index.reindex.RethrottleTests.findTaskToRethrottle(RethrottleTests.java:180)
   > 	at org.elasticsearch.index.reindex.RethrottleTests.testCase(RethrottleTests.java:94)
   > 	at org.elasticsearch.index.reindex.RethrottleTests.testDeleteByQuery(RethrottleTests.java:59)
   > 	at java.lang.Thread.run(Thread.java:748)
```

* Reproduction lines:
  * gradle :modules:reindex:test -Dtests.seed=B60C174C1CAD6D39 -Dtests.class=org.elasticsearch.index.reindex.RethrottleTests -Dtests.method="testDeleteByQueryWithWorkers" -Dtests.security.manager=true -Dtests.jvm.argline="-XX:-UseConcMarkSweepGC -XX:+UseG1GC" -Dtests.locale=id-ID -Dtests.timezone=Asia/Phnom_Penh (does not reproduce locally)
  * gradle :modules:reindex:test -Dtests.seed=B60C174C1CAD6D39 -Dtests.class=org.elasticsearch.index.reindex.RethrottleTests -Dtests.method="testDeleteByQuery" -Dtests.security.manager=true -Dtests.jvm.argline="-XX:-UseConcMarkSweepGC -XX:+UseG1GC" -Dtests.locale=id-ID -Dtests.timezone=Asia/Phnom_Penh (does not reproduce locally)

### PR DESCRIPTION
**Pull request by andyb-elastic**
When slices is set as auto, there's an additional network call
needed for the reindex tasks to know how to rethrottle. Sometimes
the rethrottle action happens before the reindex task is fully
initialized, so in the test we wait for the task to be ready.

This commit also allows reindex tasks to be able to handle
a cancel operation before they are initialized.

Closes #26192

### CODE REVIEW

**Comment by andy-elastic** [comment]
Jenkins test this please

**Comment by andy-elastic** [comment]
I'm fairly confident this fixes the issue. Some tests with larger numbers of slices seemed to have problems with the tasks not processing fast enough, but when I lowered/removed the sleep times I added (to reproduce the original issue) they seemed to go away.

File: core/src/main/java/org/elasticsearch/index/reindex/BulkByScrollTask.java [hunk 1]
```diff
@@ -170,12 +173,9 @@ public WorkerBulkByScrollTaskState getWorkerState() {
 
     @Override
     public void onCancelled() {
-        if (isLeader()) {
-            // The task cancellation task automatically finds children and cancels them, nothing extra to do
-        } else if (isWorker()) {
+        // The task cancellation task automatically finds children and cancels them, nothing extra to do for leaders
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/index/reindex/BulkByScrollTask.java]: It'd be nice to also mention that tasks that do not yet know if they are leaders or followers also have nothing to do as they can wait.

File: modules/reindex/src/test/java/org/elasticsearch/index/reindex/RethrottleTests.java [hunk 2]
```diff
@@ -174,6 +171,39 @@ private void testCase(AbstractBulkByScrollRequestBuilder<?, ?> request, String a
                 response.getBatches(), greaterThanOrEqualTo(numSlices));
     }
 
+    private ListTasksResponse rethrottleTask(TaskId taskToRethrottle, float newRequestsPerSecond) {
+        // the task isn't ready to be rethrottled until it has figured out how many slices it will use. if we rethrottle when the task is
+        // in this state, the request will fail. so we try a few times
+        long start = System.nanoTime();
+        List<Throwable> failures = new ArrayList<>();
+        do {
```
Comment (nik9000) [inline modules/reindex/src/test/java/org/elasticsearch/index/reindex/RethrottleTests.java]: I believe this is pretty similar to `assertBusy()`

Comment (andy-elastic) [inline modules/reindex/src/test/java/org/elasticsearch/index/reindex/RethrottleTests.java]: It is pretty similar; I used it at first but it only catches AssertionError, and the throwable that causes this problem is an ElasticsearchException. Think it's worth changing assertBusy or adding a similar method to ESTestCase?

Comment (nik9000) [inline modules/reindex/src/test/java/org/elasticsearch/index/reindex/RethrottleTests.java]: I've done:
```
assertBusy(() -> {
  try {
    stuff
  } catch (ElasticsearchException e) {
    throw new AssertionFailure("rethrowing to cause retry");
  }
});
```

It isn't always clearer, especially if you want to return something from the results of `assertBusy`, but I think if you don't return something then it probably is clearer. Probably.

Comment (andy-elastic) [inline modules/reindex/src/test/java/org/elasticsearch/index/reindex/RethrottleTests.java]: Right, returning something was the other thing that didn't play well. Before I had it set an AtomicReference which worked but looked a little weird

**Comment by andy-elastic** [comment]
@nik9000 I used `assertBusy` in the way you suggested, works great

**Review by nik9000 — APPROVED** [review]

**Comment by andy-elastic** [comment]
Jenkins retest this please

**Pull request merged**